pytorch_lightning/callbacks/progress/tqdm_progress.py

```diff
@@ -310,8 +319,8 @@ def print(
             s = sep.join(map(str, args))
             active_progress_bar.write(s, end=end, file=file, nolock=nolock)
 
-    def _should_update(self, current, total) -> bool:
-        return self.is_enabled and (current % self.refresh_rate == 0 or current == total)
```

Comment:
I'm not yet fully understanding this fix. How are you handling the case where we need to update the bar at the end, when total is reached? the refresh rate may not align with the last batch.

Reply:
Is that why you implement `on_test_epoch_end` now, to cover that case? In that case I agree.

Reply:
So before, each `on_*_batch_end` checked whether it was the last batch and updated in that case (by passing `total` to `_should_update`)

Now, we do it in `on_*_epoch_end` always so this is not necessary

Reply:
I see, I have tests for that so this should be fine!

Reply:
Yes, the number of calls are the same, but the logger connector state won't be the correct one in the last `on_train_batch_end` vs `on_train_epoch_end`